test/unit/cache_bin.c

```diff
@@ -379,8 +387,159 @@ TEST_BEGIN(test_cache_bin_stash) {
 }
 TEST_END
 
+typedef struct {
+	void *ptr;
+	bool is_write;
+} prefetch_arg_t;
+static prefetch_arg_t prefetch_calls[NPREFETCH_CACHE_FILL_MAX];
+static unsigned nprefetch_calls;
+
+static void
+prefetch_hook(void *p, bool is_write) {
+	prefetch_calls[nprefetch_calls].ptr = p;
+	prefetch_calls[nprefetch_calls].is_write = is_write;
+	++nprefetch_calls;
+}
+
+static void
+reset_prefetch_calls(void) {
+	nprefetch_calls = 0;
+	cache_bin_prefetch_hook_set(prefetch_hook);
+}
+
+TEST_BEGIN(test_cache_bin_array_prefetch_cl) {
+	test_skip_if(CACHELINE < sizeof(void *) * 4);
+	assert(opt_nprefetch_cache_fill == 0);
+
+	CACHE_BIN_PTR_ARRAY_DECLARE(arr, 16);
+	void *storage[16];
+	arr.ptr = storage;
+	arr.n = 16;
+
+	reset_prefetch_calls();
+	opt_nprefetch_cache_fill = 0;
+	void *fake_address = (void **)0x8badf00d;
+	storage[0] = fake_address;
+	cache_bin_ptr_array_prefetch(&arr, 1);
+	expect_zu_eq(nprefetch_calls, 0, "No calls when opt_nprefetch_cache_fill== 0");
+
+	cache_bin_prefetch_hook_set(NULL);
+}
+TEST_END
+
+
+TEST_BEGIN(test_cache_bin_alloc_easy_prefetch_disabled) {
+	/* Only makes sense if compiled with experimental fast prefetch support */
+	test_skip_if(experimental_fast_prefetch_enabled);
+
+	const int ncached_max = 10;
+	bool success;
+	void *ptr;
+
+	cache_bin_info_t info;
+	cache_bin_info_init(&info, ncached_max);
+	cache_bin_t bin;
+	test_bin_init(&bin, &info);
+
+	/* Initialize to empty; should then have 0 elements. */
+	expect_d_eq(ncached_max, cache_bin_ncached_max_get(&bin), "");
+	expect_true(cache_bin_ncached_get_local(&bin) == 0, "");
+
+	/*
+	 * We allocate fully, so we can test
+	 * prefetch at the end of the cache bin.
+	 */
+	void **ptrs = mallocx(sizeof(void *) * (ncached_max + 1), 0);
+	assert_ptr_not_null(ptrs, "Unexpected mallocx failure");
+	for  (cache_bin_sz_t i = 0; i < ncached_max; i++) {
+		expect_true(cache_bin_ncached_get_local(&bin) == i, "");
+		success = cache_bin_dalloc_easy(&bin, &ptrs[i]);
+	}
+	expect_true(cache_bin_ncached_get_local(&bin) == ncached_max,
+	    "");
+
+	reset_prefetch_calls();
+	for  (cache_bin_sz_t i = 0; i < ncached_max; i++) {
+		ptr = cache_bin_alloc_easy(&bin, &success);
+		expect_true(success, "");
+		expect_ptr_eq(ptr, &ptrs[ncached_max - i - 1], "");
+	}
+	/* Check prefetch calls */
+	expect_zu_eq(nprefetch_calls, 0, "No calls when prefetch disabled");
+
+	free(ptrs);
+	cache_bin_prefetch_hook_set(NULL);
+}
+TEST_END
+
+TEST_BEGIN(test_cache_bin_alloc_easy_prefetch_enabled) {
+	/* Only makes sense when experimental prefetch support enabled */
+	test_skip_if(!experimental_fast_prefetch_enabled);
+#ifdef CACHE_BIN_PREFETCH_STEP
+	const int step = CACHE_BIN_PREFETCH_STEP;
+#else
+	/* Does not matter as test is skipped */
+	const int step = 12345;
```

Comment:
maybe set a large-enough magic number and do assertions below to make sure `CACHE_BIN_PREFETCH_STEP` is always defined when `experimental_fast_prefetch_enabled` is true.

Reply:
Do you think we can also assert that `step` is always within a certain range instead of `!= MAGIC_STEP`? While this value is not tunable by users, such an assertion is mainly to prevent the value from being defined to a surprisingly large invalid value.